### ISSUE TITLE
Add support for dt.timezone/dt.tzinfo objects as types in pydantic models.

### ISSUE THREAD

**Issue by ollz272** [pydantic/pydantic#8895]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Would it be possible to support having tzinfo/timezone objects be used in pydantic models?

IE we have a use case where we have a timezone attached to objects in the database. We'd like to have a pydantic model that validates and represents this:

import datetime as dt

```python
class MyModel(BaseModel):
    timezone: dt.timezone | dt.tzinfo
```

This doesn't seem to be supported as of yet, but a simple conversion to a string for serialisation seems ideal. 

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
@ollz272,

This is a reasonable feature request! PRs welcome with added support for this!

**Comment by ollz272** [comment]
Great, thanks @sydney-runkle 🙌🏻 

Heres a hacked together implementation in python:
```python
from datetime import tzinfo, datetime, UTC
from typing import Type, Any

from pydantic import GetCoreSchemaHandler, BaseModel
from pydantic_core import core_schema


class TzInfo(tzinfo):
    @classmethod
    def __get_pydantic_core_schema__(cls, source: Type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
        return core_schema.no_info_wrap_validator_function(cls._validate, core_schema.any_schema())

    @classmethod
    def _validate(cls, value: Any, handler: core_schema.ValidatorFunctionWrapHandler) -> Any:
        # if we are passed an existing instance, pass it straight through.
        if isinstance(value, tzinfo):
            return str(value)

        raise ValueError("Not a valid tzinfo object")


class MyModel(BaseModel):
    name: str
    timezone: TzInfo


tz = datetime.now(tz=UTC).tzinfo
print(MyModel.model_validate({"name": "hello", "timezone": tz}))
```

I guess there would be a couple of questions here:
1. Would we go for a pure python implementation, or would some work need to be done on pydantic core?
2. Does this even belong here? Or should this live in `pydantic-extra-types`

Happy to continue some work on this

**Comment by sydney-runkle** [comment]
@ollz272,

Nice work!

> Would we go for a pure python implementation, or would some work need to be done on pydantic core?

Good question. It depends how we move forward from here. I think we should get @davidhewitt's input. I believe we might be able to keep things purely in `pydantic`, but we'll probably want to have some more complex validation / serialization logic than that used above.

> Does this even belong here? Or should this live in pydantic-extra-types

Given the mainstream nature of these types, we can keep this in `pydantic` 😄

**Comment by davidhewitt** [comment]
For sure I think `dt.timezone` can be handled directly in `pydantic`.

For `dt.tzinfo`, validation from JSON probably would create `pydantic_core.TzInfo` objects. That makes me think that handling the `tzinfo` ABC needs to go in core, but I'm not totally sure.

**Comment by davidhewitt** [comment]
It might be worth splitting the work on this into two, as those types probably behave quite differently.

**Comment by geospackle** [comment]
This looks interesting. Regarding validation from JSON, I can think of doing it with `pytz`, but we would rather have that logic in `pydantic_core` and validate against a list of valid timezones?

If that makes sense, would be happy to work on implementing validation for `tzinfo`.

**Comment by sydney-runkle** [comment]
@geospackle,

That makes sense to me re the `pydantic_core` validation, but I think we should double check with @davidhewitt before giving you the go ahead (just don't want to waste your time). Pinging him now :).

**Comment by davidhewitt** [comment]
Is `zoneinfo` preferred these days over `pytz`? We should also make sure we understand how this interacts with the `pydantic_core.TzInfo` type.

**Comment by geospackle** [comment]
@sydney-runkle 

Unclear to me if I should proceed with this.

**Comment by davidhewitt** [comment]
I think it would be difficult to interact with `pydantic_core.TzInfo` from Python, so this may need to be `pydantic_core`.

**Comment by geospackle** [comment]
> I think it would be difficult to interact with `pydantic_core.TzInfo` from Python, so this may need to be `pydantic_core`.

Makes sense. I started tinkering with this and hoping to get a draft PR up soon.

**Comment by geospackle** [comment]
This change requires more extensive work than I expected and looks like I am not able to dedicate the time needed.

It was great to learn the code base a bit better though!

**Comment by Youssefares** [comment]
Is this up for grabs? 🙂 I've done some adjacent work in [#9896](https://github.com/pydantic/pydantic/pull/9896)

**Comment by hdimer** [comment]
I'd like to take the `dt.timezone` half of this — the pure-Python part @davidhewitt flagged as doable directly in pydantic, splitting off the `dt.tzinfo`/core piece. I'll mirror the `ZoneInfo` support from #9896: a plain-validator schema with string round-trip and a JSON-schema `format`. @Youssefares, are you working on this, or ok for me to pick it up? Opening a PR shortly.

**Comment by hdimer** [comment]
Heads up: I opened #13499 with the implementation before realizing the assignment bot requires me to be assigned first, so it auto-closed. Could a maintainer assign me here? I'll reopen once assigned. It mirrors the `ZoneInfo` support and is scoped to `dt.timezone` per @davidhewitt's split (the `tzinfo` ABC left for core), with tests and docs.
